pkg/controller/common/api/v2/common.go

```diff
@@ -0,0 +1,147 @@
+package v2
+
+import (
+	"fmt"
+	"strings"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GroupVersionKind is metacontroller wrapper around schema.GroupVersionKind
+// implementing encoding.TextMarshaler and encoding.TextUnmarshaler
+type GroupVersionKind struct {
+	schema.GroupVersionKind
+}
+
+// MarshalText is implementation of  encoding.TextMarshaler
+func (gvk GroupVersionKind) MarshalText() ([]byte, error) {
+	var marshalledText string
+	if gvk.Group == "" {
+		marshalledText = fmt.Sprintf("%s.%s", gvk.Kind, gvk.Version)
+	} else {
+		marshalledText = fmt.Sprintf("%s.%s/%s", gvk.Kind, gvk.Group, gvk.Version)
+	}
+	return []byte(marshalledText), nil
+}
+
+// UnmarshalText is implementation of encoding.TextUnmarshaler
+func (gvk *GroupVersionKind) UnmarshalText(text []byte) error {
+	kindGroupVersionString := string(text)
+	parts := strings.SplitN(kindGroupVersionString, ".", 2)
+	if len(parts) < 2 {
+		return fmt.Errorf("could not unmarshall [%s], expected string in 'kind.group/version' format", string(text))
+	}
+	groupVersion, err := schema.ParseGroupVersion(parts[1])
+	if err != nil {
+		return err
+	}
+	*gvk = GroupVersionKind{
+		groupVersion.WithKind(parts[0]),
+	}
+	return nil
+}
+
+// UniformObjectMap holds objects grouped by GroupVersionKind and uniform location in the form namespace/name (if object is clustered scope, just name)
+// it holds entries in form [GroupVersionKind] -> [namespace/name] -> object
+// where namespace is empty in case of resources which are cluster scoped,
+// i.e. :
+// "v1/Pod -> 'prometheus/Prometheus'" (for namespaced child resource if the parent is cluster scope)
```

Comment:
nit: can this be improved?
e.g. Is below correct / better?
```sh
v1/Pod -> some-namespace/some-name -> Obj
```

Reply:
yes, thanks